### PR TITLE
[fix] #2465: Reimplement sumeragi node as singlethreaded state machine

### PR DESCRIPTION
**Pull request by SamHSmith**
Signed-off-by: Sam H. Smith <sam.henning.smith@protonmail.com>

### Description of the Change

Sumeragi now runs on it's own thread and is a relatively simple statemachine. To the eye it might seem more complicated because it is laid out flat. But that is to make explicit everything that is done. The final form of the code is not this. But it is a very malleable intermediate stage. Which is why it's going on the staging branch.

### Issue

#2465

### Benefits

It's actually very possible to debug this implementation as opposed to the previous version.

### Possible Drawbacks
None
### Usage Examples or Tests *[optional]*
The best way to get an understanding for the code is to run
`cargo test integration::unstable_network::unstable_network_4_peers_1_fault -- --nocapture`.
I have left in prints to stdout that should help you follow the execution.

### CODE REVIEW

File: cli/src/lib.rs
Comment (s8sato) [inline cli/src/lib.rs:301]: I'm commenting in the code so that this conversation is threaded, resolved, and collapsed.

- You consider `iroha2-staging` as a staging branch for `iroha2-dev`, right?
- You could enable CI tests by, for example, renaming `iroha2-dev` in `I2::Dev::Tests` workflow
- `iroha2-staging` is outdated relative to `iroha2-dev`. When do you plan to update?

Comment (appetrosyan) [reply cli/src/lib.rs:301]: 1. Tests should be enabled in CI, because we want more than one person to contribute to this branch. 
2. We want to merge at least once every two weeks from `iroha2-dev`. Fortunately nothing concerns the new functionality, but you need to be aware which PRs might, so occasionally have a look.

File: p2p/src/network.rs
Comment (appetrosyan) [inline p2p/src/network.rs:201]: Style point, I usually prefer having a matching `else`.

File: core/src/sumeragi/network_topology.rs
Comment (appetrosyan) [inline core/src/sumeragi/network_topology.rs:232]: So is this not supposed to be checked into `stating` or `dev`?

File: p2p/src/network.rs
Comment (appetrosyan) [inline p2p/src/network.rs:329]: Good!

File: p2p/src/network.rs
Comment (appetrosyan) [inline p2p/src/network.rs:330]:
```suggestion
        self.peers.keys().iter().cloned().collect();
```
The `iter()` is sometimes redundant, but I like to keep it.

File: cli/src/torii/routing.rs
Comment (appetrosyan) [inline cli/src/torii/routing.rs:102]: Does this definitely mean that the function definitely works?

File: cli/src/torii/routing.rs
Comment (appetrosyan) [inline cli/src/torii/routing.rs:391]: Naming convention

```suggestion
    let status = Status::from(&sumeragi.wsv_clone().metrics);
```

Or even 


```suggestion
    let status = Status::from(&sumeragi.wsv().clone().metrics);
```

And even better, following the example of our FFI function, 
```suggestion
let status = sumeragi.update_metrics(network).clone().into()
```
i.e.  `update_metrics` returns `&Metrics`, which you can clone.

File: cli/src/torii/routing.rs
Comment (appetrosyan) [inline cli/src/torii/routing.rs:392]:
```suggestion
    println!("{status:?}");
```

File: core/src/sumeragi/mod.rs
Comment (appetrosyan) [inline core/src/sumeragi/mod.rs:222]:
```suggestion
            .expect("Lock on update online peers") = online_peers;
```

File: core/src/sumeragi/mod.rs
Comment (appetrosyan) [inline core/src/sumeragi/mod.rs:169]:
```suggestion
            .expect("Get network topology lock")
```

File: core/src/sumeragi/mod.rs
Comment (appetrosyan) [inline core/src/sumeragi/mod.rs:131]: Consider returning as a recoverable error.

File: core/src/sumeragi/mod.rs
Comment (appetrosyan) [inline core/src/sumeragi/mod.rs:116]: From experience, whenever a Rust programmer sees a `Mutex`, they are itching to optimise it to an `RwLock`. Given they see `atomic` they might get rid of the `Mutex`. 

If it's a bad idea, you should explain it in a comment.

File: core/src/sumeragi/mod.rs
Comment (appetrosyan) [inline core/src/sumeragi/mod.rs:119]:
```suggestion
                incoming_message_sender: Mutex::new(incoming_message_sender),
                incoming_message_receiver: Mutex::new(incoming_message_receiver),
```

**Review by appetrosyan — CHANGES_REQUESTED** [review]

File: core/src/sumeragi/mod.rs
Comment (appetrosyan) [inline core/src/sumeragi/mod.rs:87]:
```suggestion
        let (incoming_message_sender, incoming_message_receiver) = std::sync::mpsc::channel();
```

File: core/src/sumeragi/fault.rs
Comment (appetrosyan) [inline core/src/sumeragi/fault.rs:88]: It has the word `Atomic` in it. But nothing so far suggests to me that it is `Atomic`. Moreover, a struct made up of atomic units isn't itself atomic. I know that this is WIP, but we need 0day good naming conventions.

File: core/src/sumeragi/fault.rs
Comment (appetrosyan) [inline core/src/sumeragi/fault.rs:84]:
```rust 
use std::sync::mpsc
```
```suggestion
    pub incomming_message_sender: Mutex<mpsc::Sender<Message>>,
```

File: core/src/sumeragi/fault.rs
Comment (appetrosyan) [inline core/src/sumeragi/fault.rs:98]: Why do we need this, and not just an `RwLock`, if all we're going to do is `clone` the `Vec`?

**Review by appetrosyan — CHANGES_REQUESTED** [review]

File: core/src/sumeragi/network_topology.rs
Comment (SamHSmith) [reply core/src/sumeragi/network_topology.rs:232]: Not to dev. Originally it was a note to me to remember to reimplement view changes. But instead I decided that this can live on the staging branch and there is a issue tracking it. https://github.com/hyperledger/iroha/issues/2561

File: cli/src/torii/routing.rs
Comment (SamHSmith) [reply cli/src/torii/routing.rs:102]: Yes. Because the function is only reading and the broadcast queue's and arc's get cloned properly as well. If this function didn't work then we wouldn't have 90+ % pass rate on the integration tests.

File: core/src/sumeragi/mod.rs
Comment (SamHSmith) [reply core/src/sumeragi/mod.rs:131]: good idea

File: core/src/sumeragi/fault.rs
Comment (SamHSmith) [reply core/src/sumeragi/fault.rs:98]: Because holding locks is more complicated then not doing so. `current_online_peers_by_public_key` is not a big array so this is not a problem. Cloning all the time causes performance problems. Sharing data all the time also causes performance problems and *slows down development* because there is more to think about. My guiding principle is to avoid holding locks when I don't have to. Makes things easier.

File: core/src/sumeragi/message.rs
Comment (SamHSmith) [reply core/src/sumeragi/message.rs:198]: I forgot about these since nothing in message.rs is ever called. Will remove them.

File: core/src/sumeragi/message.rs
Comment (SamHSmith) [reply core/src/sumeragi/message.rs]: I'm using println because logging output is disable for some reason in the test `integration::unstable_network::unstable_network_4_peers_1_fault`. Of course they will be removed or replaced by logging before we merge into iroha2-dev. But for now they stay.

File: core/src/sumeragi/mod.rs
Comment (SamHSmith) [reply core/src/sumeragi/mod.rs:169]: yes

File: core/src/sumeragi/mod.rs
Comment (SamHSmith) [reply core/src/sumeragi/mod.rs:131]: Turns out, actually not a good idea. Mutex poison errors can't be passed between threads safely.

File: core/src/sumeragi/mod.rs
Comment (SamHSmith) [reply core/src/sumeragi/mod.rs:119]: thx

File: p2p/src/network.rs
Comment (SamHSmith) [reply p2p/src/network.rs:201]: I like it like this when it is a pure early return. As opposed to a different code path.

File: p2p/src/network.rs
Comment (SamHSmith) [reply p2p/src/network.rs:330]: yes

File: cli/src/torii/routing.rs
Comment (SamHSmith) [reply cli/src/torii/routing.rs:391]: using suggestion 1

File: cli/src/torii/routing.rs
Comment (SamHSmith) [reply cli/src/torii/routing.rs:392]: that seems redundant

File: core/src/sumeragi/fault.rs
Comment (Erigara) [inline core/src/sumeragi/fault.rs:650]: `zip` could be used here.

File: core/src/sumeragi/fault.rs
Comment (Erigara) [inline core/src/sumeragi/fault.rs:674]: I think `filter_map` would be convinient here.

**Review by Erigara — COMMENTED** [review]
I found current state good starting point for further improvements.

File: core/src/sumeragi/fault.rs
Comment (outoftardis) [inline core/src/sumeragi/fault.rs:379]: I'd remove the `absolutely not acceptable` remark, or at least rephrase it somehow

File: core/src/sumeragi/network_topology.rs
Comment (SamHSmith) [reply core/src/sumeragi/network_topology.rs:232]: yes

File: core/src/sumeragi/fault.rs
Comment (SamHSmith) [reply core/src/sumeragi/fault.rs:84]: sure

File: core/src/sumeragi/fault.rs
Comment (SamHSmith) [reply core/src/sumeragi/fault.rs:88]: I've changed it to SumeragiStateMachineData. Good of you two for pushing me on this.

File: core/src/sumeragi/fault.rs
Comment (SamHSmith) [reply core/src/sumeragi/fault.rs:98]: Yes. And allowing concurrent reading access makes programmers think it's okay to hold the lock. Because it avoids a clone and "that's good right?". In this case it is very important that the data is able to be updated. I'd rather avoid the bug of "sometimes the online peers won't update" that got introduced because someone thought they were being smart. You guys being so resistant to the mutex proves why it is good. It communicates very strongly. DO NOT HOLD THIS LOCK. DOING SO WILL BE BAD. A RWLock communicates a more seducive "Look at how smart I'm being by allowing multiple readers on the same data."

A codebase where one is very hesitant to hold locks leads to a more healthy view on multithreading. It's fast when threads don't communicate. When they have independent work to do. And the only reason for communicating between threads is to arrange for more work to be done in solitude.

File: cli/src/lib.rs
Comment (SamHSmith) [reply cli/src/lib.rs:78]: Rust supports imports inside functions. The reason being that you want to keep your imports as close to the place they are used as possible, if they are only used by that one piece of code.

There has to be a relay, otherwise there's a cyclical dependency between p2p and sumeragi. My instinct was to merge the two modules but since that would have been very destructive I chose this method instead.

File: core/src/sumeragi/mod.rs
Comment (SamHSmith) [reply core/src/sumeragi/mod.rs:127]: sure

File: core/src/sumeragi/fault.rs
Comment (SamHSmith) [reply core/src/sumeragi/fault.rs:129]: The inlining seems premature but otherwise yes.

File: core/src/sumeragi/fault.rs
Comment (SamHSmith) [reply core/src/sumeragi/fault.rs:198]: It isn't.

File: core/src/sumeragi/fault.rs
Comment (SamHSmith) [reply core/src/sumeragi/fault.rs:358]: We don't *need* all these scopes. But it was constructed this way because the scopes loosely reflect what would previously been function calls. These should be refined over time. But they communicate things about what is connected to what.

Comment (Erigara) [reply core/src/sumeragi/fault.rs:358]: Ok, thanks you, for clarification!

File: cli/src/lib.rs
Comment (s8sato) [inline cli/src/lib.rs:301]: Failed tests as of 0390bd29 compared to 03ef82b1:

- [ ] `integration::asset_propagation::client_add_asset_quantity_to_existing_asset_should_increase_asset_amount_on_another_peer`
- [ ] `integration::connected_peers::connected_peers_with_f_1_0_1`
- [ ] `integration::connected_peers::connected_peers_with_f_2_1_2`
- [ ] `integration::events::pipeline::transaction_with_fail_instruction_should_be_rejected`
- [ ] `integration::events::pipeline::transaction_with_no_instructions_should_be_committed`
- [ ] `integration::multisignature_transaction::multisignature_transactions_should_wait_for_all_signatures`
- [ ] `integration::unregister_peer::unstable_network_stable_after_add_and_after_remove_peer`

If it's intentional, I'd like an explanation

**Review by s8sato — COMMENTED** [review]
Let me take a closer look at the main loop

File: cli/src/lib.rs
Comment (SamHSmith) [reply cli/src/lib.rs:301]: In order to allow multiple people to work on the reimplementation, this PR has things that are broken. Then we will fix these with more PR's into iroha2-staging. We only merge into iroha2-dev when the reimplementation is complete and everything works. Aleksandr for instance is working on #2561 which fixes most of the tests you mention. And if some are still failing after that we will fix them.

File: core/src/sumeragi/mod.rs
Comment (SamHSmith) [reply core/src/sumeragi/mod.rs:116]: I will do so.

File: core/src/sumeragi/fault.rs
Comment (SamHSmith) [reply core/src/sumeragi/fault.rs:165]: There doesn't seem to be an issue currently. It was implemented though in the previous implementation. This is a, "we'll see if we end up needing this", sort of situation.

File: core/src/sumeragi/fault.rs
Comment (SamHSmith) [reply core/src/sumeragi/fault.rs:379]: yea

File: core/src/sumeragi/fault.rs
Comment (SamHSmith) [reply core/src/sumeragi/fault.rs:674]: It was indeed useful.

**Review by appetrosyan — APPROVED** [review]

**Review by s8sato — APPROVED** [review]

**Pull request merged**
